compiler/frontend/src/org/jetbrains/kotlin/resolve/DescriptorResolver.java

```diff
@@ -327,12 +327,12 @@ else if (containingDescriptor instanceof FunctionDescriptor || containingDescrip
         return defaultVisibility;
     }
 
-    public static Modality getDefaultModality(DeclarationDescriptor containingDescriptor, boolean isBodyPresent) {
+    public static Modality getDefaultModality(DeclarationDescriptor containingDescriptor, Visibility visibility, boolean isBodyPresent) {
         Modality defaultModality;
         if (containingDescriptor instanceof ClassDescriptor) {
             boolean isTrait = ((ClassDescriptor) containingDescriptor).getKind() == ClassKind.TRAIT;
             boolean isDefinitelyAbstract = isTrait && !isBodyPresent;
-            Modality basicModality = isTrait ? Modality.OPEN : Modality.FINAL;
+            Modality basicModality = isTrait && visibility != Visibilities.PRIVATE ? Modality.OPEN : Modality.FINAL;
```

Comment:
Please use `Visibilities.isPrivate` instead & add a test on a private-to-this fun/val in a trait